### ISSUE TITLE
HTTP client return type `Map` incorrect

### ISSUE THREAD

**Issue by saschagrunert** [angular/angular#31610]
# 🐞 bug report

### Affected Package

The issue is caused by package @angular/common/http

### Is this a regression?

I'm unsure.

### Description

Related to this discussion: https://github.com/kubernetes-sigs/release-notes/pull/75#discussion_r304444918

If I do an `http.get` with a return type of `Map<...>`, then the actual returned type is `any`, like this

```typescript
    export type NoteList = Map<number, Note>;

    return this.http.get<NoteList>('assets/data/file.json').pipe(
      map((notes: NoteList) => {
        console.log(notes.keys());
        return notes;
      }),
    );
```

This results in a type error (typescript 3.4.5).

```
TypeError: notes.keys is not a function
```

## 🔬 Minimal Reproduction

Simply try the source code of https://github.com/kubernetes-sigs/release-notes

## 🌍  Your Environment

**Angular Version:**
<pre><code>

> ./node_modules/.bin/ng --version

     _                      _                 ____ _     ___
    / \   _ __   __ _ _   _| | __ _ _ __     / ___| |   |_ _|
   / △ \ | '_ \ / _` | | | | |/ _` | '__|   | |   | |    | |
  / ___ \| | | | (_| | |_| | | (_| | |      | |___| |___ | |
 /_/   \_\_| |_|\__, |\__,_|_|\__,_|_|       \____|_____|___|
                |___/

Angular CLI: 8.1.0
Node: 10.16.0
OS: linux x64
Angular: 8.1.0
... animations, cli, common, compiler, compiler-cli, core, forms
... language-service, platform-browser, platform-browser-dynamic
... router

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.801.0
@angular-devkit/build-angular     0.801.0
@angular-devkit/build-optimizer   0.801.0
@angular-devkit/build-webpack     0.801.0
@angular-devkit/core              8.1.0
@angular-devkit/schematics        8.1.0
@ngtools/webpack                  8.1.0
@schematics/angular               8.1.0
@schematics/update                0.801.0
rxjs                              6.5.2
typescript                        3.4.5
webpack                           4.35.2
</code></pre>

**Anything else relevant?**

I don't think so, no.

**Comment by Airblader** [comment]
Typescript cannot actually convert any data (remember, it's a compile time language), the generic is purely a type assertion of what you expect to get, which is probably just a json object since you can't possibly get a Map from an HTTP call.

You need to use the correct type assertion and convert the data into a Map yourself.

**Comment by saschagrunert** [comment]
> You need to convert the data into a Map yourself if you want to work on a Map instance. The generic on the get call, if you want to use it at all, should only describe the format of the data as received by the http call.

Yeah, but why does the `HttpClient` have such an API if the return type is always an `any`? It does not make much sense to me...

**Comment by Airblader** [comment]
The return type is not `any`, the return type is what you specify as the generic. I think you're confusing the compile-time TypeScript type with the runtime JavaScript type. Those are very different things.

The generic exists solely for convenience because writing `http.get<Foo>()` is easier than having to write `http.get().pipe(map(foo => foo as Foo))` just to type it in the following flow.
